### DISCUSSION TITLE
For NoData alerts, we need a way to set it apart from "For".

### DISCUSSION THREAD

**Discussion by yheonhochoi** [grafana/grafana#46156] [Alerting]
**What would you like to be added**:

In a Grafana alerting configuration, it would be nice if the NoData alerting can be set separately from the duration `For` of other alerting settings.

**Why is this needed**:

In a typical alerting configuration, the `For` value indicates how long the metric met the threshold, but in the case of NoData, it is set to whether the data is available, unlike other alerts.

There is confusion because the meaning of two alerts is set to a single value.

For example, we might want to set the alerts triggered by the threshold to short durations, but the NoData alerts may want to make a longer setting for reasons such as collection delay.

If the settings of the two alerts are separated, it is likely to be more flexible.

**Comment by kylebrandt** [comment]
Currently FOR, which causes the pending state, only happens for the Alerting state and not the NoData state (https://github.com/grafana/grafana/issues/16290). Currently exploring making this work for NoData in https://github.com/grafana/grafana/pull/18119 .

Having separate FOR settings for different "Bad States" does propose a logic complication. Consider the following case:

Configuration:
- The alert is configured with a FOR for NoData for 1 minute
- The alert is configured with a FOR for Alerting for 2 minutes
  
Data Changes:
- The alert is Ok,  then:
- Loop: The alert is NoData for 30 seconds, The alert is Alerting for 1 minute, GOTO Loop

In this case, neither FOR condition has been meant (assuming the FOR resets on a state change). This means in the scenario a notification would not be sent for either NoData or Alerting even though the thing is "bad".

So I think to handle different FORs and still not miss the above scenario, the state logic will be more complex than just having a single shared FOR condition. I think it is probably better to keep it simple even though the separate conditions are more flexible.

**Comment by patricktokeeffe** [comment]
I also see value in this use case. However I believe allowing multiple alert definitions would be a better time investment than complicating the alert logic.

**Comment by marefr** [comment]
Related #18940

**Comment by bljordan** [comment]
I'm ambivalent to whether or not NoData uses the same FOR setting as the alert, but I would like to see NoData respect some duration setting before firing. Please see the closed referenced issue above for the specific scenario.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). I'm closing feature requests that have seen limited from the community and/or the Grafana team. 

Should this feature still be of interest to you, please re-create it in our [Discussions area](https://github.com/grafana/grafana/discussions). In doing so the community can continue to contribute to exploring, confirming and sharing their use cases (and more).

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Should you wish to connect to discuss anything Grafana-related, I have an open door policy and [my contact details are on my GitHub profile page](https://github.com/pkolyvas).

**Comment by gotjosh** [comment]
@grobinson-grafana, @yuri-tceretian and I have been discussing the possibility of including this in some shape or form. 

Turning this into a feature request to see how much support/traction we can get for this feature.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82989. Feel free to continue the discussion around this there. Thank you!
